Add `-Z proc-macro-backtrace` to allow showing proc-macro panics

Aaron1011 opened the pull request:
Fixes #75050

Previously, we would unconditionally suppress the panic hook during
proc-macro execution. This commit adds a new flag
`-Z proc-macro-backtrace`, which allows running the panic hook for
easier debugging.

rust-highfive commented:
r? @petrochenkov

(rust_highfive has picked a reviewer for you, use r? to override)

petrochenkov commented:
(I'll review this next weekend.)

petrochenkov commented:
I'll review this next weekend.
I'm not sure this is something that needs to be done, or at least the priority is not high.

bors commented:
:umbrella: The latest upstream changes (presumably #74862) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
r=me after rebase

Aaron1011 committed:
Add `-Z proc-macro-backtrace` to allow showing proc-macro panics

Fixes #75050

Previously, we would unconditionally suppress the panic hook during
proc-macro execution. This commit adds a new flag
-Z proc-macro-backtrace, which allows running the panic hook for
easier debugging.

Aaron1011 commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit d9208665b5eef883c025f44717f8c76afa871a94 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov d9208665b5eef883c025f44717f8c76afa871a94 -->
<!-- homu: {"type":"Approved","sha":"d9208665b5eef883c025f44717f8c76afa871a94","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit d9208665b5eef883c025f44717f8c76afa871a94 with merge e98f0632bbec24b196dbd6fc820537f6f3724807...
<!-- homu: {"type":"BuildStarted","head_sha":"d9208665b5eef883c025f44717f8c76afa871a94","merge_sha":"e98f0632bbec24b196dbd6fc820537f6f3724807"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1049501125), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36322)
Approved by: petrochenkov
Pushing e98f0632bbec24b196dbd6fc820537f6f3724807 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36322","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1049501125"},"merge_sha":"e98f0632bbec24b196dbd6fc820537f6f3724807"} -->

rust-log-analyzer commented:
Your PR [failed](https://github.com/rust-lang-ci/rust/runs/1049581017) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/e98f0632bbec24b196dbd6fc820537f6f3724807/checks/1049581017/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/e98f0632bbec24b196dbd6fc820537f6f3724807/checks/1049581017/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))
